### PR TITLE
Send active-ack after log collection for nonblocking activations.

### PR DESCRIPTION
**Pull request by cbickel**
Until now, an active-ack is sent before logs of a container are collected. If one customer writes a lot of logs or if log-collection is slow for some other reason, the invoker already gets new activations, that are queueing up.
This PR changes the behavior, to send the active-ack (for non-blocking) activations after log collection is finished. For blocking actions, there are two active acks now. One with the response for the user and one to free up the space in the bookkeeping of the loadbalancer.

I've put this proposal on the dev-list for discussion: https://lists.apache.org/thread.html/726c802f38f3872d057f9adb6a52043eeb8fbd68b601c57bdb12d706@%3Cdev.openwhisk.apache.org%3E

I'll continue work on this PR, if we come to a conclusion how to proceed with this issue.

## Related issue and scope
- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [x] Loadbalancer
- [x] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [x] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by rabbah** [comment]
This is fine for non blocking. For blocking did you consider a second message (resource release in addition to active ack)?

-r

**Comment by cbickel** [comment]
@rabbah Using a new type of message, that is smaller, definitely makes sense.

File: core/controller/src/main/scala/whisk/core/loadBalancer/ShardingContainerPoolBalancer.scala [hunk 1]
```diff
@@ -284,7 +284,12 @@ class ShardingContainerPoolBalancer(config: WhiskConfig, controllerInstance: Con
     activations.getOrElseUpdate(
       msg.activationId, {
         val timeoutHandler = actorSystem.scheduler.scheduleOnce(timeout) {
-          processCompletion(Left(msg.activationId), msg.transid, forced = true, invoker = instance)
+          processCompletion(
+            Left(msg.activationId),
+            msg.transid,
+            forced = true,
+            invoker = instance,
+            isLogCollectionFinished = true)
```
Comment (dubee) [inline core/controller/src/main/scala/whisk/core/loadBalancer/ShardingContainerPoolBalancer.scala]: Good find on this one, @cbickel!

It looks like `isLogCollectionFinished` is set to `true` for forced active acks. Can you describe the intended behavior for forced active acks with the new log flag?

**Comment by cbickel** [comment]
I'll change the code and reopen the PR afterwards.

**Comment by cbickel** [comment]
PG3#2926 looks good

**Comment by dgrove-oss** [comment]
I think this change is fine in and of itself, but it caused downstream problems for kube-deploy.  https://github.com/apache/incubator-openwhisk-deploy-kube/issues/328.  Documenting here in case it impacts other downstream deployment projects like Mesos.

**Pull request merged**